pkg/build/controller/image_change_controller.go

```diff
@@ -78,12 +76,18 @@ func (c *ImageChangeController) HandleImageRepo() {
 			glog.V(4).Infof("Running build for buildConfig %s in namespace %s", config.Name, config.Namespace)
 			b := buildutil.GenerateBuildFromConfig(config, nil, imageSubstitutions)
 			if err := c.BuildCreator.Create(config.Namespace, b); err != nil {
-				glog.V(2).Infof("Error starting build for buildConfig %v: %v", config.Name, err)
+				return fmt.Errorf("Error starting build for buildConfig %v: %v, will be retried", config.Name, err)
```

Comment:
the controller is the bit that has the logic to know whether a particular event can/should be retried or not (based on what state has been modified or would be duplicated if the event is replayed) so i don't see how you can(well, should) defer the decision to retry out of the controller.  The fact is the controller is making the decision to retry or not based on the type of error it returns and knowing the logic of the retry controller anyway.  There's really no avoiding the fact that the controller needs to have some knowledge of the types of errors the retry handler is expecting and how it will handle them.  the less of that there is, the better, imho.  (which is why I preferred the approach that just returned RetryableError vs FatalError wrapping the real error)